source/core/shape.cc

```diff
@@ -311,6 +311,12 @@ Tensor<1, dim>
 Sphere<dim>::gradient(const Point<dim> &evaluation_point,
                       const unsigned int /*component*/) const
 {
+  // We make sure that the evaluation point and the sphere center are different,
+  // because if they are the same the analytical gradient is not defined: the
+  // function returns a NaN. We use the numerical gradient if the points are the
+  // same.
+  if ((evaluation_point - this->position).norm() < 1e-12)
```

Comment:
You might want to make the scaling of this < operator depend on the radius of the sphere no? To be coherent with dimensions.
Otherwise all good.